// @restatedev/restate-sdk-core/dist/serde_api.cjs

//#region src/serde_api.ts
var BinarySerde = class {
	contentType = "application/octet-stream";
	serialize(value) {
		return value;
	}
	deserialize(data) {
		return data;
	}
};
var VoidSerde = class {
	serialize(value) {
		if (value !== void 0) throw new Error("Expected undefined value");
		return new Uint8Array(0);
	}
	deserialize(data) {
		if (data.length !== 0) throw new Error("Expected empty data");
	}
};
var StandardSchemaSerde = class {
	contentType = "application/json";
	jsonSchema;
	constructor(schema, validateOptions, jsonSchemaOptions) {
		this.schema = schema;
		this.validateOptions = validateOptions;
		const standard = schema["~standard"];
		if (isStandardJSONSchemaV1(standard)) try {
			this.jsonSchema = standard.jsonSchema.output({
				target: "draft-2020-12",
				libraryOptions: jsonSchemaOptions
			});
		} catch {
			this.jsonSchema = void 0;
		}
		const testResult = standard.validate(void 0);
		if (testResult && typeof testResult === "object" && "then" in testResult) {} else if (!testResult.issues) this.contentType = void 0;
	}
	serialize(value) {
		if (value === void 0) return new Uint8Array(0);
		return new TextEncoder().encode(JSON.stringify(value));
	}
	deserialize(data) {
		const js = data.length === 0 ? void 0 : JSON.parse(new TextDecoder().decode(data));
		const result = this.schema["~standard"].validate(js, this.validateOptions);
		if (result && typeof result === "object" && "then" in result) throw new TypeError("Async validation is not supported in Serde. Restate SDK supports only synchronous validation.");
		if (result.issues) {
			const errorMessages = result.issues.map(formatStandardSchemaIssue).join("\n");
			throw new TypeError(`Standard schema validation failed:\n${errorMessages}`);
		}
		return result.value;
	}
};
function formatStandardSchemaIssue(issue) {
	if (issue.path && issue.path.length > 0) return `* (at ${"/" + issue.path.map((p) => {
		if (typeof p === "object" && p !== null && "key" in p) return String(p.key);
		return String(p);
	}).join("/")}) ${issue.message}`;
	return `* ${issue.message}`;
}
function isStandardJSONSchemaV1(standard) {
	return standard != void 0 && "jsonSchema" in standard && typeof standard.jsonSchema === "object" && standard.jsonSchema !== null && "output" in standard.jsonSchema && typeof standard.jsonSchema.output === "function";
}
let serde;
(function(_serde) {
	class JsonSerde {
		contentType = "application/json";
		constructor(jsonSchema) {
			this.jsonSchema = jsonSchema;
		}
		serialize(value) {
			if (value === void 0) return new Uint8Array(0);
			return new TextEncoder().encode(JSON.stringify(value));
		}
		deserialize(data) {
			if (data.length === 0) return;
			return JSON.parse(new TextDecoder().decode(data));
		}
		schema(schema) {
			return new JsonSerde(schema);
		}
	}
	_serde.JsonSerde = JsonSerde;
	_serde.json = new JsonSerde();
	_serde.binary = new BinarySerde();
	_serde.empty = new VoidSerde();
	_serde.schema = (schema, validateOptions, jsonSchemaOptions) => {
		return new StandardSchemaSerde(schema, validateOptions, jsonSchemaOptions);
	};
})(serde || (serde = {}));

//#endregion
Object.defineProperty(exports, 'serde', {
  enumerable: true,
  get: function () {
    return serde;
  }
});
