// @restatedev/restate-sdk-core/dist/serde_api.js
/*
 * Copyright (c) 2023-2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate SDK for Node.js/TypeScript,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in file LICENSE in the root
 * directory of this repository or package, or at
 * https://github.com/restatedev/sdk-typescript/blob/main/LICENSE
 */
class BinarySerde {
    contentType = "application/octet-stream";
    serialize(value) {
        return value;
    }
    deserialize(data) {
        return data;
    }
}
class VoidSerde {
    serialize(value) {
        if (value !== undefined) {
            throw new Error("Expected undefined value");
        }
        return new Uint8Array(0);
    }
    deserialize(data) {
        if (data.length !== 0) {
            throw new Error("Expected empty data");
        }
    }
}
class StandardSchemaSerde {
    schema;
    validateOptions;
    contentType = "application/json";
    jsonSchema;
    constructor(schema, validateOptions, jsonSchemaOptions) {
        this.schema = schema;
        this.validateOptions = validateOptions;
        // Extract JSON schema if available
        const standard = schema["~standard"];
        if (isStandardJSONSchemaV1(standard)) {
            try {
                this.jsonSchema = standard.jsonSchema.output({
                    target: "draft-2020-12",
                    libraryOptions: jsonSchemaOptions,
                });
            }
            catch {
                // If JSON schema generation fails, leave it undefined
                this.jsonSchema = undefined;
            }
        }
        // Check if schema is for void/undefined type
        // Standard Schema doesn't have a direct way to detect void types,
        // so we serialize undefined and check if validation succeeds
        const testResult = standard.validate(undefined);
        // Handle both sync and async validation results
        if (testResult && typeof testResult === "object" && "then" in testResult) {
            // If it's a Promise, we can't determine contentType synchronously
            // Keep contentType as "application/json"
        }
        else if (!testResult.issues) {
            this.contentType = undefined;
        }
    }
    serialize(value) {
        if (value === undefined) {
            return new Uint8Array(0);
        }
        return new TextEncoder().encode(JSON.stringify(value));
    }
    deserialize(data) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        const js = data.length === 0
            ? undefined
            : JSON.parse(new TextDecoder().decode(data));
        const result = this.schema["~standard"].validate(js, this.validateOptions);
        // Standard Schema validate can return a Promise, but Serde must be synchronous
        if (result && typeof result === "object" && "then" in result) {
            throw new TypeError("Async validation is not supported in Serde. Restate SDK supports only synchronous validation.");
        }
        if (result.issues) {
            const errorMessages = result.issues
                .map(formatStandardSchemaIssue)
                .join("\n");
            throw new TypeError(`Standard schema validation failed:\n${errorMessages}`);
        }
        return result.value;
    }
}
function formatStandardSchemaIssue(issue) {
    if (issue.path && issue.path.length > 0) {
        const jsonPointer = "/" +
            issue.path
                .map((p) => {
                // Handle both PathSegment objects and primitive PropertyKey values
                if (typeof p === "object" && p !== null && "key" in p) {
                    return String(p.key);
                }
                return String(p);
            })
                .join("/");
        return `* (at ${jsonPointer}) ${issue.message}`;
    }
    return `* ${issue.message}`;
}
function isStandardJSONSchemaV1(standard) {
    return (standard != undefined &&
        "jsonSchema" in standard &&
        typeof standard.jsonSchema === "object" &&
        standard.jsonSchema !== null &&
        "output" in standard.jsonSchema &&
        typeof standard.jsonSchema.output === "function");
}
export var serde;
(function (serde) {
    class JsonSerde {
        jsonSchema;
        contentType = "application/json";
        constructor(jsonSchema) {
            this.jsonSchema = jsonSchema;
        }
        serialize(value) {
            if (value === undefined) {
                return new Uint8Array(0);
            }
            return new TextEncoder().encode(JSON.stringify(value));
        }
        deserialize(data) {
            if (data.length === 0) {
                return undefined;
            }
            return JSON.parse(new TextDecoder().decode(data));
        }
        schema(schema) {
            return new JsonSerde(schema);
        }
    }
    serde.JsonSerde = JsonSerde;
    serde.json = new JsonSerde();
    serde.binary = new BinarySerde();
    serde.empty = new VoidSerde();
    /**
     * A Standard Schema-based serde.
     *
     * @param schema the standard schema
     * @param validateOptions options passed to `StandardSchemaV1.Options.libraryOptions` when validating
     * @param jsonSchemaOptions options passed to `StandardJsonSchemaV1.Options.libraryOptions` for code generation
     * @returns a serde that will validate the data with the standard schema
     */
    serde.schema = (schema, validateOptions, jsonSchemaOptions) => {
        return new StandardSchemaSerde(schema, validateOptions, jsonSchemaOptions);
    };
})(serde || (serde = {}));
