// @restatedev/restate-sdk/dist/utils/rand.cjs
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.cjs');
let node_buffer = require("node:buffer");
node_buffer = require_rolldown_runtime.__toESM(node_buffer);

//#region src/utils/rand.ts
var RandImpl = class RandImpl {
	randstate256;
	constructor(id, checkState = () => void 0) {
		this.checkState = checkState;
		if (typeof id === "bigint") {
			let x = id & RandImpl.U64_MASK;
			const next = () => {
				x = x + 11400714819323198485n & RandImpl.U64_MASK;
				let z = x;
				z ^= z >> 30n;
				z = z * 13787848793156543929n & RandImpl.U64_MASK;
				z ^= z >> 27n;
				z = z * 10723151780598845931n & RandImpl.U64_MASK;
				z ^= z >> 31n;
				return z & RandImpl.U64_MASK;
			};
			this.randstate256 = [
				next(),
				next(),
				next(),
				next()
			];
		} else this.randstate256 = id;
	}
	static U64_MASK = (1n << 64n) - 1n;
	u64() {
		const result = RandImpl.rotl(this.randstate256[0] + this.randstate256[3] & RandImpl.U64_MASK, 23n) + this.randstate256[0] & RandImpl.U64_MASK;
		const t = this.randstate256[1] << 17n & RandImpl.U64_MASK;
		this.randstate256[2] ^= this.randstate256[0];
		this.randstate256[3] ^= this.randstate256[1];
		this.randstate256[1] ^= this.randstate256[2];
		this.randstate256[0] ^= this.randstate256[3];
		this.randstate256[2] ^= t;
		this.randstate256[3] = RandImpl.rotl(this.randstate256[3], 45n);
		return result;
	}
	static rotl(x, k) {
		return x << k & RandImpl.U64_MASK | x >> 64n - k;
	}
	static U53_MASK = (1n << 53n) - 1n;
	random() {
		this.checkState("rand.random");
		const u53 = this.u64() & RandImpl.U53_MASK;
		return Number(u53) / 2 ** 53;
	}
	uuidv4() {
		this.checkState("rand.uuidv4");
		const buf = node_buffer.Buffer.alloc(16);
		buf.writeBigUInt64LE(this.u64(), 0);
		buf.writeBigUInt64LE(this.u64(), 8);
		buf[6] = buf[6] & 15 | 64;
		buf[8] = buf[8] & 63 | 128;
		return uuidStringify(buf);
	}
};
const byteToHex = [];
for (let i = 0; i < 256; ++i) byteToHex.push((i + 256).toString(16).slice(1));
/**
* Convert array of 16 byte values to UUID string format of the form:
* XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
*/
function uuidStringify(arr, offset = 0) {
	return (byteToHex[arr[offset + 0]] + byteToHex[arr[offset + 1]] + byteToHex[arr[offset + 2]] + byteToHex[arr[offset + 3]] + "-" + byteToHex[arr[offset + 4]] + byteToHex[arr[offset + 5]] + "-" + byteToHex[arr[offset + 6]] + byteToHex[arr[offset + 7]] + "-" + byteToHex[arr[offset + 8]] + byteToHex[arr[offset + 9]] + "-" + byteToHex[arr[offset + 10]] + byteToHex[arr[offset + 11]] + byteToHex[arr[offset + 12]] + byteToHex[arr[offset + 13]] + byteToHex[arr[offset + 14]] + byteToHex[arr[offset + 15]]).toLowerCase();
}

//#endregion
exports.RandImpl = RandImpl;
