// @restatedev/restate-sdk/dist/endpoint/components.js
import { HandlerKind } from "../types/rpc.js";
import { millisOrDurationToMillis, serde } from "@restatedev/restate-sdk-core";

//#region src/endpoint/components.ts
function handlerInputDiscovery(handler, defaultSerde) {
	const serde$1 = handler.inputSerde ?? defaultSerde;
	let contentType = void 0;
	let jsonSchema = void 0;
	if (serde$1.jsonSchema) {
		jsonSchema = serde$1.jsonSchema;
		contentType = handler.accept ?? serde$1.contentType;
	} else if (handler.accept) contentType = handler.accept;
	else if (serde$1.contentType) contentType = serde$1.contentType;
	else return {};
	return {
		required: false,
		contentType,
		jsonSchema
	};
}
function handlerOutputDiscovery(handler, defaultSerde) {
	const serde$1 = handler.outputSerde ?? defaultSerde;
	let contentType = void 0;
	let jsonSchema = void 0;
	if (serde$1.jsonSchema) {
		jsonSchema = serde$1.jsonSchema;
		contentType = serde$1.contentType ?? "application/json";
	} else if (serde$1.contentType) contentType = serde$1.contentType;
	else return { setContentTypeIfEmpty: false };
	return {
		setContentTypeIfEmpty: false,
		jsonSchema,
		contentType
	};
}
var ServiceComponent = class {
	handlers = /* @__PURE__ */ new Map();
	constructor(componentName, description, metadata, options) {
		this.componentName = componentName;
		this.description = description;
		this.metadata = metadata;
		this.options = options;
	}
	name() {
		return this.componentName;
	}
	add(name, handlerWrapper) {
		this.handlers.set(name, new ServiceHandler$1(name, handlerWrapper, this));
	}
	discovery() {
		const handlers = [...this.handlers.entries()].map(([name, handler]) => {
			return {
				name,
				...commonHandlerOptions(handler.handlerWrapper, this.options?.serde ?? serde.json)
			};
		});
		return {
			name: this.componentName,
			ty: "SERVICE",
			handlers,
			documentation: this.description,
			metadata: this.metadata,
			...commonServiceOptions(this.options)
		};
	}
	handlerMatching(url) {
		return this.handlers.get(url.handlerName);
	}
};
var ServiceHandler$1 = class {
	constructor(handlerName, handlerWrapper, parent) {
		this.handlerName = handlerName;
		this.handlerWrapper = handlerWrapper;
		this.parent = parent;
	}
	name() {
		return this.handlerName;
	}
	component() {
		return this.parent;
	}
	kind() {
		return this.handlerWrapper.kind;
	}
	invoke(context, input) {
		return this.handlerWrapper.invoke(context, input);
	}
};
var VirtualObjectComponent = class {
	handlers = /* @__PURE__ */ new Map();
	constructor(componentName, description, metadata, options) {
		this.componentName = componentName;
		this.description = description;
		this.metadata = metadata;
		this.options = options;
	}
	name() {
		return this.componentName;
	}
	add(name, wrapper) {
		this.handlers.set(name, new VirtualObjectHandler(name, wrapper, this));
	}
	discovery() {
		const handlers = [...this.handlers.entries()].map(([name, handler]) => {
			return {
				name,
				ty: handler.kind() === HandlerKind.EXCLUSIVE ? "EXCLUSIVE" : "SHARED",
				...commonHandlerOptions(handler.handlerWrapper, this.options?.serde ?? serde.json)
			};
		});
		return {
			name: this.componentName,
			ty: "VIRTUAL_OBJECT",
			handlers,
			documentation: this.description,
			metadata: this.metadata,
			...commonServiceOptions(this.options)
		};
	}
	handlerMatching(url) {
		return this.handlers.get(url.handlerName);
	}
};
var VirtualObjectHandler = class {
	constructor(handlerName, handlerWrapper, parent) {
		this.handlerName = handlerName;
		this.handlerWrapper = handlerWrapper;
		this.parent = parent;
	}
	name() {
		return this.handlerName;
	}
	component() {
		return this.parent;
	}
	kind() {
		return this.handlerWrapper.kind;
	}
	invoke(context, input) {
		return this.handlerWrapper.invoke(context, input);
	}
};
var WorkflowComponent = class {
	handlers = /* @__PURE__ */ new Map();
	constructor(componentName, description, metadata, options) {
		this.componentName = componentName;
		this.description = description;
		this.metadata = metadata;
		this.options = options;
	}
	name() {
		return this.componentName;
	}
	add(name, wrapper) {
		this.handlers.set(name, new WorkflowHandler$1(name, wrapper, this));
	}
	discovery() {
		const handlers = [...this.handlers.entries()].map(([name, handler]) => {
			return {
				name,
				ty: handler.kind() === HandlerKind.WORKFLOW ? "WORKFLOW" : "SHARED",
				workflowCompletionRetention: handler.kind() === HandlerKind.WORKFLOW && this.options?.workflowRetention !== void 0 ? millisOrDurationToMillis(this.options?.workflowRetention) : void 0,
				...commonHandlerOptions(handler.handlerWrapper, this.options?.serde ?? serde.json)
			};
		});
		return {
			name: this.componentName,
			ty: "WORKFLOW",
			handlers,
			documentation: this.description,
			metadata: this.metadata,
			...commonServiceOptions(this.options)
		};
	}
	handlerMatching(url) {
		return this.handlers.get(url.handlerName);
	}
};
var WorkflowHandler$1 = class {
	constructor(handlerName, handlerWrapper, parent) {
		this.handlerName = handlerName;
		this.handlerWrapper = handlerWrapper;
		this.parent = parent;
	}
	name() {
		return this.handlerName;
	}
	component() {
		return this.parent;
	}
	kind() {
		return this.handlerWrapper.kind;
	}
	invoke(context, input) {
		return this.handlerWrapper.invoke(context, input);
	}
};
function parseUrlComponents(urlPath) {
	if (!urlPath) return {
		type: "unknown",
		path: ""
	};
	const fragments = urlPath.split("/");
	if (fragments.length >= 3 && fragments[fragments.length - 3] === "invoke") return {
		type: "invoke",
		componentName: fragments[fragments.length - 2],
		handlerName: fragments[fragments.length - 1]
	};
	if (fragments.length > 0 && fragments[fragments.length - 1] === "discover") return { type: "discover" };
	if (fragments.length > 0 && fragments[fragments.length - 1] === "health") return { type: "health" };
	return {
		type: "unknown",
		path: urlPath
	};
}
function commonServiceOptions(options) {
	return {
		journalRetention: options?.journalRetention !== void 0 ? millisOrDurationToMillis(options.journalRetention) : void 0,
		idempotencyRetention: options?.idempotencyRetention !== void 0 ? millisOrDurationToMillis(options.idempotencyRetention) : void 0,
		inactivityTimeout: options?.inactivityTimeout !== void 0 ? millisOrDurationToMillis(options.inactivityTimeout) : void 0,
		abortTimeout: options?.abortTimeout !== void 0 ? millisOrDurationToMillis(options.abortTimeout) : void 0,
		ingressPrivate: options?.ingressPrivate,
		enableLazyState: options !== void 0 && "enableLazyState" in options ? options.enableLazyState : void 0,
		retryPolicyExponentiationFactor: options?.retryPolicy?.exponentiationFactor,
		retryPolicyInitialInterval: options?.retryPolicy?.initialInterval !== void 0 ? millisOrDurationToMillis(options?.retryPolicy?.initialInterval) : void 0,
		retryPolicyMaxInterval: options?.retryPolicy?.maxInterval !== void 0 ? millisOrDurationToMillis(options?.retryPolicy?.maxInterval) : void 0,
		retryPolicyMaxAttempts: options?.retryPolicy?.maxAttempts,
		retryPolicyOnMaxAttempts: options?.retryPolicy?.onMaxAttempts === "kill" ? "KILL" : options?.retryPolicy?.onMaxAttempts === "pause" ? "PAUSE" : void 0
	};
}
function commonHandlerOptions(wrapper, defaultSerde) {
	return {
		input: handlerInputDiscovery(wrapper, defaultSerde),
		output: handlerOutputDiscovery(wrapper, defaultSerde),
		journalRetention: wrapper.journalRetention !== void 0 ? millisOrDurationToMillis(wrapper.journalRetention) : void 0,
		idempotencyRetention: wrapper.idempotencyRetention !== void 0 ? millisOrDurationToMillis(wrapper.idempotencyRetention) : void 0,
		inactivityTimeout: wrapper.inactivityTimeout !== void 0 ? millisOrDurationToMillis(wrapper.inactivityTimeout) : void 0,
		abortTimeout: wrapper.abortTimeout !== void 0 ? millisOrDurationToMillis(wrapper.abortTimeout) : void 0,
		ingressPrivate: wrapper.ingressPrivate,
		enableLazyState: wrapper.enableLazyState,
		retryPolicyExponentiationFactor: wrapper.retryPolicy?.exponentiationFactor,
		retryPolicyInitialInterval: wrapper.retryPolicy?.initialInterval !== void 0 ? millisOrDurationToMillis(wrapper.retryPolicy?.initialInterval) : void 0,
		retryPolicyMaxInterval: wrapper.retryPolicy?.maxInterval !== void 0 ? millisOrDurationToMillis(wrapper.retryPolicy?.maxInterval) : void 0,
		retryPolicyMaxAttempts: wrapper.retryPolicy?.maxAttempts,
		retryPolicyOnMaxAttempts: wrapper.retryPolicy?.onMaxAttempts === "kill" ? "KILL" : wrapper.retryPolicy?.onMaxAttempts === "pause" ? "PAUSE" : void 0,
		documentation: wrapper.description,
		metadata: wrapper.metadata
	};
}

//#endregion
export { ServiceComponent, VirtualObjectComponent, WorkflowComponent, parseUrlComponents };
