// @restatedev/restate-sdk/dist/endpoint/node_endpoint.cjs
const require_rolldown_runtime = require('../_virtual/rolldown_runtime.cjs');
const require_errors = require('../types/errors.cjs');
const require_generic = require('./handlers/generic.cjs');
const require_endpoint = require('./endpoint.cjs');
let http2 = require("http2");
http2 = require_rolldown_runtime.__toESM(http2);
let node_stream = require("node:stream");
node_stream = require_rolldown_runtime.__toESM(node_stream);

//#region src/endpoint/node_endpoint.ts
var NodeEndpoint = class {
	builder = new require_endpoint.EndpointBuilder();
	bind(definition) {
		this.builder.bind(definition);
		return this;
	}
	withIdentityV1(...keys) {
		this.builder.addIdentityKeys(...keys);
		return this;
	}
	defaultServiceOptions(options) {
		this.builder.setDefaultServiceOptions(options);
		return this;
	}
	setLogger(logger) {
		this.builder.setLogger(logger);
		return this;
	}
	journalValueCodecProvider(codecProvider) {
		this.builder.setJournalValueCodecProvider(codecProvider);
		return this;
	}
	http2Handler() {
		return nodeHttp2Handler(this.builder.build());
	}
	listen(port) {
		const endpoint = this.builder.build();
		const actualPort = port ?? parseInt(process.env.PORT ?? "9080");
		endpoint.rlog.info(`Restate SDK started listening on ${actualPort}...`);
		const server = http2.createServer(nodeHttp2Handler(endpoint));
		return new Promise((resolve, reject) => {
			let failed = false;
			server.once("error", (e) => {
				failed = true;
				reject(e);
			});
			server.listen(actualPort, () => {
				if (failed) return;
				const address = server.address();
				if (address === null || typeof address === "string") reject(/* @__PURE__ */ new TypeError("endpoint.listen() currently supports only binding to a PORT"));
				else resolve(address.port);
			});
		});
	}
};
function nodeHttp2Handler(endpoint) {
	const handler = new require_generic.GenericHandler(endpoint, "BIDI_STREAM", {});
	return (request, response) => {
		(async () => {
			const abortController = new AbortController();
			request.once("aborted", () => {
				abortController.abort();
			});
			request.once("close", () => {
				abortController.abort();
			});
			request.once("error", () => {
				abortController.abort();
			});
			if (request.destroyed || request.aborted) {
				endpoint.rlog.error("Client disconnected");
				abortController.abort();
			}
			try {
				const url = request.url;
				const webRequestBody = node_stream.Readable.toWeb(request);
				const resp = await handler.handle({
					url,
					headers: request.headers,
					body: webRequestBody,
					extraArgs: [],
					abortSignal: abortController.signal
				});
				if (response.destroyed) return;
				response.writeHead(resp.statusCode, resp.headers);
				const responseWeb = node_stream.Writable.toWeb(response);
				await resp.body.pipeTo(responseWeb);
			} catch (e) {
				const error = require_errors.ensureError(e);
				const logger = require_generic.tryCreateContextualLogger(endpoint.loggerTransport, request.url, request.headers) ?? endpoint.rlog;
				if (error.name === "AbortError") logger.error("Got abort error from connection: " + error.message + "\nThis might indicate that:\n* The restate-server aborted the connection after hitting the 'abort-timeout'\n* The connection with the restate-server was lost\n\nPlease check the invocation in the Restate UI for more details.");
				else logger.error("Error while handling request: " + (error.stack ?? error.message));
				response.destroy(error);
				abortController.abort();
			}
		})().catch(() => {});
	};
}

//#endregion
exports.NodeEndpoint = NodeEndpoint;
