// @restatedev/restate-sdk/dist/common_api.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const require_errors = require('./types/errors.cjs');
const require_rpc = require('./types/rpc.cjs');
const require_context = require('./context.cjs');
const require_internal = require('./internal.cjs');
let __restatedev_restate_sdk_core = require("@restatedev/restate-sdk-core");
__restatedev_restate_sdk_core = require_rolldown_runtime.__toESM(__restatedev_restate_sdk_core);

//#region src/common_api.ts
/**
* @deprecated Use {@link RestatePromise}
*/
const CombineablePromise = require_context.RestatePromise;
const createServiceHandler = require_rpc.handlers.handler;
const createObjectHandler = require_rpc.handlers.object.exclusive;
const createObjectSharedHandler = require_rpc.handlers.object.shared;
const createWorkflowHandler = require_rpc.handlers.workflow.workflow;
const createWorkflowSharedHandler = require_rpc.handlers.workflow.shared;

//#endregion
exports.CombineablePromise = CombineablePromise;
exports.createObjectHandler = createObjectHandler;
exports.createObjectSharedHandler = createObjectSharedHandler;
exports.createServiceHandler = createServiceHandler;
exports.createWorkflowHandler = createWorkflowHandler;
exports.createWorkflowSharedHandler = createWorkflowSharedHandler;
