// @restatedev/restate-sdk/dist/context_impl.js
import { INTERNAL_ERROR_CODE, RestateError, RetryableError, TerminalError, UNKNOWN_ERROR_CODE, ensureError, logError } from "./types/errors.js";
import { HandlerKind, makeRpcCallProxy, makeRpcSendProxy } from "./types/rpc.js";
import { CompletablePromise } from "./utils/completable_promise.js";
import { WasmCommandType, WasmHeader } from "./endpoint/handlers/vm/sdk_shared_core_wasm_bindings.js";
import { RandImpl } from "./utils/rand.js";
import { InvocationPendingPromise, PromisesExecutor, RestateCombinatorPromise, RestateInvocationPromise, RestatePendingPromise, RestateSinglePromise, extractContext, pendingPromise } from "./promises.js";
import { InputPump, OutputPump } from "./io.js";
import { millisOrDurationToMillis, serde } from "@restatedev/restate-sdk-core";

//#region src/context_impl.ts
var ContextImpl = class {
	rand;
	date = {
		now: () => {
			return this.run(() => Date.now());
		},
		toJSON: () => {
			return this.run(() => (/* @__PURE__ */ new Date()).toJSON());
		}
	};
	outputPump;
	runClosuresTracker;
	promisesExecutor;
	defaultSerde;
	constructor(coreVm, input, console, handlerKind, vmLogger, invocationRequest, invocationEndPromise, inputReader, outputWriter, journalValueCodec, defaultSerde, asTerminalError) {
		this.coreVm = coreVm;
		this.input = input;
		this.console = console;
		this.handlerKind = handlerKind;
		this.vmLogger = vmLogger;
		this.invocationRequest = invocationRequest;
		this.invocationEndPromise = invocationEndPromise;
		this.journalValueCodec = journalValueCodec;
		this.asTerminalError = asTerminalError;
		this.rand = new RandImpl(input.random_seed, () => {});
		this.outputPump = new OutputPump(coreVm, outputWriter);
		this.runClosuresTracker = new RunClosuresTracker();
		this.promisesExecutor = new PromisesExecutor(coreVm, new InputPump(coreVm, inputReader, this.handleInvocationEndError.bind(this)), this.outputPump, this.runClosuresTracker, this.promiseExecutorErrorCallback.bind(this));
		this.defaultSerde = defaultSerde ?? serde.json;
	}
	isProcessing() {
		return this.coreVm.is_processing();
	}
	cancel(invocationId) {
		this.processNonCompletableEntry(WasmCommandType.CancelInvocation, () => {}, (vm) => vm.sys_cancel_invocation(invocationId));
	}
	attach(invocationId, serde$1) {
		return this.processCompletableEntry(WasmCommandType.AttachInvocation, () => {}, (vm) => vm.sys_attach_invocation(invocationId), SuccessWithSerde(serde$1 ?? this.defaultSerde, this.journalValueCodec), Failure);
	}
	get key() {
		switch (this.handlerKind) {
			case HandlerKind.EXCLUSIVE:
			case HandlerKind.SHARED:
			case HandlerKind.WORKFLOW: return this.input.key;
			default: throw new TerminalError("this handler type doesn't support key()");
		}
	}
	request() {
		return this.invocationRequest;
	}
	get(name, serde$1) {
		return this.processCompletableEntry(WasmCommandType.GetState, () => {}, (vm) => vm.sys_get_state(name), VoidAsNull, SuccessWithSerde(serde$1 ?? this.defaultSerde, this.journalValueCodec));
	}
	stateKeys() {
		return this.processCompletableEntry(WasmCommandType.GetStateKeys, () => {}, (vm) => vm.sys_get_state_keys(), StateKeys);
	}
	set(name, value, serde$1) {
		this.processNonCompletableEntry(WasmCommandType.SetState, () => this.journalValueCodec.encode((serde$1 ?? this.defaultSerde).serialize(value)), (vm, bytes) => vm.sys_set_state(name, bytes));
	}
	clear(name) {
		this.processNonCompletableEntry(WasmCommandType.ClearState, () => {}, (vm) => vm.sys_clear_state(name));
	}
	clearAll() {
		this.processNonCompletableEntry(WasmCommandType.ClearAllState, () => {}, (vm) => vm.sys_clear_all_state());
	}
	genericCall(call) {
		const requestSerde = call.inputSerde ?? serde.binary;
		const responseSerde = call.outputSerde ?? serde.binary;
		let parameter;
		try {
			parameter = this.journalValueCodec.encode(requestSerde.serialize(call.parameter));
		} catch (e) {
			this.handleInvocationEndError(e, (vm, error) => vm.notify_error_for_next_command(error.message, error.stack, WasmCommandType.Call));
			return new InvocationPendingPromise(this);
		}
		try {
			const call_handles = this.coreVm.sys_call(call.service, call.method, parameter, call.key, call.headers ? Object.entries(call.headers).map(([key, value]) => new WasmHeader(key, value)) : [], call.idempotencyKey, call.name);
			const commandIndex = this.coreVm.last_command_index();
			const invocationIdPromise = new RestateSinglePromise(this, call_handles.invocation_id_completion_id, completeCommandPromiseUsing(WasmCommandType.Call, commandIndex, InvocationIdCompleter));
			return new RestateInvocationPromise(this, call_handles.call_completion_id, completeCommandPromiseUsing(WasmCommandType.Call, commandIndex, SuccessWithSerde(responseSerde, this.journalValueCodec), Failure), invocationIdPromise);
		} catch (e) {
			this.handleInvocationEndError(e);
			return new InvocationPendingPromise(this);
		}
	}
	genericSend(send) {
		const requestSerde = send.inputSerde ?? serde.binary;
		let parameter;
		try {
			parameter = this.journalValueCodec.encode(requestSerde.serialize(send.parameter));
		} catch (e) {
			this.handleInvocationEndError(e, (vm, error) => vm.notify_error_for_next_command(error.message, error.stack, WasmCommandType.OneWayCall));
			return new InvocationPendingPromise(this);
		}
		try {
			const delay = send.delay !== void 0 ? millisOrDurationToMillis(send.delay) : void 0;
			const handles = this.coreVm.sys_send(send.service, send.method, parameter, send.key, send.headers ? Object.entries(send.headers).map(([key, value]) => new WasmHeader(key, value)) : [], delay !== void 0 && delay > 0 ? BigInt(delay) : void 0, send.idempotencyKey, send.name);
			const commandIndex = this.coreVm.last_command_index();
			return { invocationId: new RestateSinglePromise(this, handles.invocation_id_completion_id, completeCommandPromiseUsing(WasmCommandType.OneWayCall, commandIndex, InvocationIdCompleter)) };
		} catch (e) {
			this.handleInvocationEndError(e);
			return { invocationId: pendingPromise() };
		}
	}
	serviceClient({ name }) {
		return makeRpcCallProxy((call) => this.genericCall(call), this.defaultSerde, name);
	}
	objectClient({ name }, key) {
		return makeRpcCallProxy((call) => this.genericCall(call), this.defaultSerde, name, key);
	}
	workflowClient({ name }, key) {
		return makeRpcCallProxy((call) => this.genericCall(call), this.defaultSerde, name, key);
	}
	serviceSendClient({ name }, opts) {
		return makeRpcSendProxy((send) => this.genericSend(send), this.defaultSerde, name, void 0, opts?.delay);
	}
	objectSendClient({ name }, key, opts) {
		return makeRpcSendProxy((send) => this.genericSend(send), this.defaultSerde, name, key, opts?.delay);
	}
	workflowSendClient({ name }, key, opts) {
		return makeRpcSendProxy((send) => this.genericSend(send), this.defaultSerde, name, key, opts?.delay);
	}
	run(nameOrAction, actionSecondParameter, options) {
		const { name, action } = unpackRunParameters(nameOrAction, actionSecondParameter);
		const serde$1 = options?.serde ?? this.defaultSerde;
		let handle;
		try {
			handle = this.coreVm.sys_run(name ?? "");
		} catch (e) {
			this.handleInvocationEndError(e);
			return new RestatePendingPromise(this);
		}
		const commandIndex = this.coreVm.last_command_index();
		const doRun = async () => {
			const startTime = Date.now();
			let res;
			let err;
			try {
				res = await action();
			} catch (e) {
				err = ensureError(e, this.asTerminalError);
			}
			const attemptDuration = Date.now() - startTime;
			try {
				if (err !== void 0) if (err instanceof TerminalError) this.coreVm.propose_run_completion_failure(handle, {
					code: err.code,
					message: err.message,
					metadata: []
				});
				else if (err instanceof RetryableError) {
					const maxRetryDuration = options?.maxRetryDuration ?? options?.maxRetryDurationMillis;
					this.coreVm.propose_run_completion_failure_transient_with_delay_override(handle, err.message, err.stack, BigInt(attemptDuration), err.retryAfter !== void 0 ? BigInt(millisOrDurationToMillis(err.retryAfter)) : void 0, options?.maxRetryAttempts, maxRetryDuration !== void 0 ? BigInt(millisOrDurationToMillis(maxRetryDuration)) : void 0);
				} else {
					this.vmLogger.warn(`Error when processing ctx.run '${name}'.\n`, err);
					let retryPolicy;
					if (options?.retryIntervalFactor !== void 0 || options?.maxRetryAttempts !== void 0 || options?.initialRetryInterval !== void 0 || options?.initialRetryIntervalMillis !== void 0 || options?.maxRetryDuration !== void 0 || options?.maxRetryDurationMillis !== void 0 || options?.maxRetryInterval !== void 0 || options?.maxRetryIntervalMillis !== void 0) {
						const maxRetryDuration = options?.maxRetryDuration ?? options?.maxRetryDurationMillis;
						retryPolicy = {
							factor: options?.retryIntervalFactor ?? 2,
							initial_interval: millisOrDurationToMillis(options?.initialRetryInterval ?? options?.initialRetryIntervalMillis ?? 50),
							max_attempts: options?.maxRetryAttempts,
							max_duration: maxRetryDuration === void 0 ? void 0 : millisOrDurationToMillis(maxRetryDuration),
							max_interval: millisOrDurationToMillis(options?.maxRetryInterval ?? options?.maxRetryIntervalMillis ?? { seconds: 10 })
						};
					}
					this.coreVm.propose_run_completion_failure_transient(handle, err.message, err.stack, BigInt(attemptDuration), retryPolicy);
				}
				else {
					const serializedRes = serde$1.serialize(res);
					const encodedRes = this.journalValueCodec.encode(serializedRes);
					this.coreVm.propose_run_completion_success(handle, encodedRes);
				}
			} catch (e) {
				this.handleInvocationEndError(e);
				return pendingPromise();
			}
			await this.outputPump.awaitNextProgress();
		};
		this.runClosuresTracker.registerRunClosure(handle, doRun);
		return new RestateSinglePromise(this, handle, completeCommandPromiseUsing(WasmCommandType.Run, commandIndex, SuccessWithSerde(serde$1, this.journalValueCodec), Failure));
	}
	sleep(duration, name) {
		return this.processCompletableEntry(WasmCommandType.Sleep, () => {
			if (duration === void 0) throw new Error(`Duration is undefined.`);
			const millis = millisOrDurationToMillis(duration);
			if (millis < 0) throw new Error(`Invalid negative sleep duration: ${millis}ms.\nIf this duration is computed from a desired wake up time, make sure to record 'now' using 'wakeUpTime - ctx.date.now()'.`);
			return BigInt(millis);
		}, (vm, millis) => vm.sys_sleep(millis, name), VoidAsUndefined);
	}
	awakeable(serde$1) {
		let awakeable;
		try {
			awakeable = this.coreVm.sys_awakeable();
		} catch (e) {
			this.handleInvocationEndError(e);
			return {
				id: "invalid",
				promise: new RestatePendingPromise(this)
			};
		}
		return {
			id: awakeable.id,
			promise: new RestateSinglePromise(this, awakeable.handle, completeSignalPromiseUsing(VoidAsUndefined, SuccessWithSerde(serde$1 ?? this.defaultSerde, this.journalValueCodec), Failure))
		};
	}
	resolveAwakeable(id, payload, serde$1) {
		this.processNonCompletableEntry(WasmCommandType.CompleteAwakeable, () => {
			let value;
			if (serde$1) value = payload === void 0 ? new Uint8Array() : serde$1.serialize(payload);
			else value = payload !== void 0 ? this.defaultSerde.serialize(payload) : this.defaultSerde.serialize(null);
			return this.journalValueCodec.encode(value);
		}, (vm, bytes) => vm.sys_complete_awakeable_success(id, bytes));
	}
	rejectAwakeable(id, reason) {
		this.processNonCompletableEntry(WasmCommandType.CompleteAwakeable, () => {}, (vm) => {
			vm.sys_complete_awakeable_failure(id, {
				code: UNKNOWN_ERROR_CODE,
				message: reason,
				metadata: []
			});
		});
	}
	promise(name, serde$1) {
		return new DurablePromiseImpl(this, name, serde$1);
	}
	static createCombinator(combinatorConstructor, promises) {
		const self = extractContext(promises[0]);
		if (!self) throw new Error("Not a combinable promise");
		const castedPromises = [];
		for (const promise of promises) {
			if (extractContext(promise) !== self) {
				self.handleInvocationEndError(/* @__PURE__ */ new Error("You're mixing up RestatePromises from different RestateContext. This is not supported."));
				return new RestatePendingPromise(self);
			}
			castedPromises.push(promise);
		}
		return new RestateCombinatorPromise(self, combinatorConstructor, castedPromises);
	}
	processNonCompletableEntry(commandType, prepare, vmCall) {
		let input;
		try {
			input = prepare();
		} catch (e) {
			this.handleInvocationEndError(e, (vm, error) => vm.notify_error_for_next_command(error.message, error.stack, commandType));
			return;
		}
		try {
			vmCall(this.coreVm, input);
		} catch (e) {
			this.handleInvocationEndError(e);
		}
	}
	processCompletableEntry(commandType, prepare, vmCall, ...completers) {
		let input;
		try {
			input = prepare();
		} catch (e) {
			this.handleInvocationEndError(e, (vm, error) => vm.notify_error_for_next_command(error.message, error.stack, commandType));
			return new RestatePendingPromise(this);
		}
		let handle;
		try {
			handle = vmCall(this.coreVm, input);
		} catch (e) {
			this.handleInvocationEndError(e);
			return new RestatePendingPromise(this);
		}
		const commandIndex = this.coreVm.last_command_index();
		return new RestateSinglePromise(this, handle, completeCommandPromiseUsing(commandType, commandIndex, ...completers));
	}
	promiseExecutorErrorCallback(e) {
		if (e instanceof AsyncCompleterError) {
			const cause = ensureError(e.cause);
			logError(this.vmLogger, e.cause);
			this.coreVm.notify_error_for_specific_command(cause.message, cause.stack, e.commandType, e.commandIndex, null);
		} else {
			const error = ensureError(e);
			logError(this.vmLogger, error);
			if (!(error instanceof RestateError)) this.coreVm.notify_error(error.message, error.stack);
		}
		this.invocationEndPromise.resolve();
	}
	handleInvocationEndError(e, notify_vm_error = (vm, error$1) => {
		vm.notify_error(error$1.message, error$1.stack);
	}) {
		const error = ensureError(e);
		logError(this.vmLogger, error);
		notify_vm_error(this.coreVm, error);
		this.invocationEndPromise.resolve();
	}
};
function unpackRunParameters(a, b) {
	if (typeof a === "string") {
		if (typeof b !== "function") throw new TypeError("");
		return {
			name: a,
			action: b
		};
	}
	if (typeof a !== "function") throw new TypeError("unexpected type at the first parameter");
	if (b) throw new TypeError("unexpected a function as a second parameter.");
	return { action: a };
}
var DurablePromiseImpl = class {
	serde;
	constructor(ctx, name, serde$1) {
		this.ctx = ctx;
		this.name = name;
		this.serde = serde$1 ?? this.ctx.defaultSerde;
	}
	then(onfulfilled, onrejected) {
		return this.get().then(onfulfilled, onrejected);
	}
	catch(onrejected) {
		return this.get().catch(onrejected);
	}
	finally(onfinally) {
		return this.get().finally(onfinally);
	}
	[Symbol.toStringTag] = "DurablePromise";
	get() {
		return this.ctx.processCompletableEntry(WasmCommandType.GetPromise, () => {}, (vm) => vm.sys_get_promise(this.name), SuccessWithSerde(this.serde, this.ctx.journalValueCodec), Failure);
	}
	peek() {
		return this.ctx.processCompletableEntry(WasmCommandType.PeekPromise, () => {}, (vm) => vm.sys_peek_promise(this.name), VoidAsUndefined, SuccessWithSerde(this.serde, this.ctx.journalValueCodec), Failure);
	}
	resolve(value) {
		return this.ctx.processCompletableEntry(WasmCommandType.CompletePromise, () => this.ctx.journalValueCodec.encode(this.serde.serialize(value)), (vm, bytes) => vm.sys_complete_promise_success(this.name, bytes), VoidAsUndefined, Failure);
	}
	reject(errorMsg) {
		return this.ctx.processCompletableEntry(WasmCommandType.CompletePromise, () => {}, (vm) => vm.sys_complete_promise_failure(this.name, {
			code: INTERNAL_ERROR_CODE,
			message: errorMsg,
			metadata: []
		}), VoidAsUndefined, Failure);
	}
};
var RunClosuresTracker = class {
	currentRunWaitPoint;
	runsToExecute = /* @__PURE__ */ new Map();
	executeRun(handle) {
		const runClosure = this.runsToExecute.get(handle);
		if (runClosure === void 0) throw new Error(`Handle ${handle} doesn't exist`);
		runClosure().finally(() => {
			this.unblockCurrentRunWaitPoint();
		}).catch(() => {});
	}
	registerRunClosure(handle, runClosure) {
		this.runsToExecute.set(handle, runClosure);
	}
	awaitNextCompletedRun() {
		if (this.currentRunWaitPoint === void 0) this.currentRunWaitPoint = new CompletablePromise();
		return this.currentRunWaitPoint.promise;
	}
	unblockCurrentRunWaitPoint() {
		if (this.currentRunWaitPoint !== void 0) {
			const p = this.currentRunWaitPoint;
			this.currentRunWaitPoint = void 0;
			p.resolve();
		}
	}
};
var AsyncCompleterError = class {
	constructor(cause, commandType, commandIndex) {
		this.cause = cause;
		this.commandType = commandType;
		this.commandIndex = commandIndex;
	}
};
function completeCommandPromiseUsing(commandType, commandIndex, ...completers) {
	return async (value, prom) => {
		try {
			for (const completer of completers) if (await completer(value, prom)) return;
		} catch (e) {
			throw new AsyncCompleterError(e, commandType, commandIndex);
		}
		throw new Error(`Unexpected variant in async result: ${JSON.stringify(value)}`);
	};
}
function completeSignalPromiseUsing(...completers) {
	return async (value, prom) => {
		for (const completer of completers) if (await completer(value, prom)) return;
		throw new Error(`Unexpected variant in async result: ${JSON.stringify(value)}`);
	};
}
const VoidAsNull = (value, prom) => {
	if (value === "Empty") {
		prom.resolve(null);
		return Promise.resolve(true);
	}
	return Promise.resolve(false);
};
const VoidAsUndefined = (value, prom) => {
	if (value === "Empty") {
		prom.resolve(void 0);
		return Promise.resolve(true);
	}
	return Promise.resolve(false);
};
function SuccessWithSerde(serde$1, journalCodec, transform) {
	return async (value, prom) => {
		if (typeof value !== "object" || !("Success" in value)) return false;
		let buffer;
		if (journalCodec !== void 0) buffer = await journalCodec.decode(value.Success);
		else buffer = value.Success;
		let val = serde$1.deserialize(buffer);
		if (transform) val = transform(val);
		prom.resolve(val);
		return true;
	};
}
const Failure = (value, prom) => {
	if (typeof value === "object" && "Failure" in value) {
		prom.reject(new TerminalError(value.Failure.message, { errorCode: value.Failure.code }));
		return Promise.resolve(true);
	}
	return Promise.resolve(false);
};
const StateKeys = (value, prom) => {
	if (typeof value === "object" && "StateKeys" in value) {
		prom.resolve(value.StateKeys);
		return Promise.resolve(true);
	}
	return Promise.resolve(false);
};
const InvocationIdCompleter = (value, prom) => {
	if (typeof value === "object" && "InvocationId" in value) {
		prom.resolve(value.InvocationId);
		return Promise.resolve(true);
	}
	return Promise.resolve(false);
};

//#endregion
export { ContextImpl };
