// @restatedev/restate-sdk/dist/context.js
import { ContextImpl } from "./context_impl.js";

//#region src/context.ts
const InvocationIdParser = { fromString(id) {
	if (!id.startsWith("inv")) throw new Error(`Expected invocation id to start with 'inv' but got ${id}`);
	return id;
} };
const RestatePromise = {
	all(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return ContextImpl.createCombinator((p) => Promise.all(p), values);
	},
	race(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return ContextImpl.createCombinator((p) => Promise.race(p), values);
	},
	any(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return ContextImpl.createCombinator((p) => Promise.any(p), values);
	},
	allSettled(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return ContextImpl.createCombinator((p) => Promise.allSettled(p), values);
	}
};

//#endregion
export { InvocationIdParser, RestatePromise };
