// @restatedev/restate-sdk/dist/context.cjs
const require_context_impl = require('./context_impl.cjs');

//#region src/context.ts
const InvocationIdParser = { fromString(id) {
	if (!id.startsWith("inv")) throw new Error(`Expected invocation id to start with 'inv' but got ${id}`);
	return id;
} };
const RestatePromise = {
	all(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return require_context_impl.ContextImpl.createCombinator((p) => Promise.all(p), values);
	},
	race(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return require_context_impl.ContextImpl.createCombinator((p) => Promise.race(p), values);
	},
	any(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return require_context_impl.ContextImpl.createCombinator((p) => Promise.any(p), values);
	},
	allSettled(values) {
		if (values.length === 0) throw new Error("Expected combineable promise to have at least one promise");
		return require_context_impl.ContextImpl.createCombinator((p) => Promise.allSettled(p), values);
	}
};

//#endregion
exports.InvocationIdParser = InvocationIdParser;
exports.RestatePromise = RestatePromise;
