// @restatedev/restate-sdk/dist/promises.js
import { CancelledError, RestateError, TerminalError, TimeoutError } from "./types/errors.js";
import { CompletablePromise } from "./utils/completable_promise.js";
import { setImmediate } from "node:timers/promises";

//#region src/promises.ts
function pendingPromise() {
	return new Promise(() => {});
}
var PromiseState = /* @__PURE__ */ function(PromiseState$1) {
	PromiseState$1[PromiseState$1["COMPLETED"] = 0] = "COMPLETED";
	PromiseState$1[PromiseState$1["NOT_COMPLETED"] = 1] = "NOT_COMPLETED";
	return PromiseState$1;
}(PromiseState || {});
const RESTATE_CTX_SYMBOL = Symbol("restateContext");
function extractContext(n) {
	return n[RESTATE_CTX_SYMBOL];
}
var AbstractRestatePromise = class {
	[RESTATE_CTX_SYMBOL];
	pollingPromise;
	cancelPromise = new CompletablePromise();
	constructor(ctx) {
		this[RESTATE_CTX_SYMBOL] = ctx;
	}
	then(onfulfilled, onrejected) {
		this.pollingPromise = this.pollingPromise || this[RESTATE_CTX_SYMBOL].promisesExecutor.doProgress(this).catch(() => {});
		return this.publicPromiseOrCancelPromise().then(onfulfilled, onrejected);
	}
	catch(onrejected) {
		this.pollingPromise = this.pollingPromise || this[RESTATE_CTX_SYMBOL].promisesExecutor.doProgress(this).catch(() => {});
		return this.publicPromiseOrCancelPromise().catch(onrejected);
	}
	finally(onfinally) {
		this.pollingPromise = this.pollingPromise || this[RESTATE_CTX_SYMBOL].promisesExecutor.doProgress(this).catch(() => {});
		return this.publicPromiseOrCancelPromise().finally(onfinally);
	}
	publicPromiseOrCancelPromise() {
		return Promise.race([this.cancelPromise.promise, this.publicPromise()]);
	}
	orTimeout(duration) {
		return new RestateCombinatorPromise(this[RESTATE_CTX_SYMBOL], ([thisPromise, sleepPromise]) => {
			return new Promise((resolve, reject) => {
				thisPromise.then(resolve, reject);
				sleepPromise.then(() => {
					reject(new TimeoutError());
				}, reject);
			});
		}, [this, this[RESTATE_CTX_SYMBOL].sleep(duration)]);
	}
	map(mapper) {
		return new RestateMappedPromise(this[RESTATE_CTX_SYMBOL], this, mapper);
	}
	tryCancel() {
		this.cancelPromise.reject(new CancelledError());
	}
};
var RestateSinglePromise = class extends AbstractRestatePromise {
	state = PromiseState.NOT_COMPLETED;
	completablePromise = new CompletablePromise();
	constructor(ctx, handle, completer) {
		super(ctx);
		this.handle = handle;
		this.completer = completer;
	}
	uncompletedLeaves() {
		return this.state === PromiseState.COMPLETED ? [] : [this.handle];
	}
	async tryComplete() {
		if (this.state === PromiseState.COMPLETED) return;
		const notification = this[RESTATE_CTX_SYMBOL].coreVm.take_notification(this.handle);
		if (notification === "NotReady") return;
		this.state = PromiseState.COMPLETED;
		await this.completer(notification, this.completablePromise);
	}
	publicPromise() {
		return this.completablePromise.promise;
	}
	[Symbol.toStringTag] = "RestateSinglePromise";
};
var RestateInvocationPromise = class extends RestateSinglePromise {
	constructor(ctx, handle, completer, invocationIdPromise) {
		super(ctx, handle, completer);
		this.invocationIdPromise = invocationIdPromise;
	}
	get invocationId() {
		return this.invocationIdPromise;
	}
};
var RestateCombinatorPromise = class extends AbstractRestatePromise {
	state = PromiseState.NOT_COMPLETED;
	combinatorPromise;
	constructor(ctx, combinatorConstructor, childs) {
		super(ctx);
		this.childs = childs;
		this.combinatorPromise = combinatorConstructor(childs.map((p) => p.publicPromise())).finally(() => {
			this.state = PromiseState.COMPLETED;
		});
	}
	uncompletedLeaves() {
		return this.state === PromiseState.COMPLETED ? [] : this.childs.flatMap((p) => p.uncompletedLeaves());
	}
	async tryComplete() {
		await Promise.allSettled(this.childs.map((c) => c.tryComplete()));
	}
	publicPromise() {
		return this.combinatorPromise;
	}
	[Symbol.toStringTag] = "RestateCombinatorPromise";
};
var RestatePendingPromise = class {
	[RESTATE_CTX_SYMBOL];
	constructor(ctx) {
		this[RESTATE_CTX_SYMBOL] = ctx;
	}
	then(onfulfilled, onrejected) {
		return pendingPromise().then(onfulfilled, onrejected);
	}
	catch(onrejected) {
		return pendingPromise().catch(onrejected);
	}
	finally(onfinally) {
		return pendingPromise().finally(onfinally);
	}
	orTimeout() {
		return this;
	}
	map() {
		return this;
	}
	tryCancel() {}
	async tryComplete() {}
	uncompletedLeaves() {
		return [];
	}
	publicPromise() {
		return pendingPromise();
	}
	[Symbol.toStringTag] = "RestatePendingPromise";
};
var InvocationPendingPromise = class extends RestatePendingPromise {
	constructor(ctx) {
		super(ctx);
	}
	get invocationId() {
		return pendingPromise();
	}
};
var RestateMappedPromise = class extends AbstractRestatePromise {
	publicPromiseMapper;
	constructor(ctx, inner, mapper) {
		super(ctx);
		this.inner = inner;
		this.publicPromiseMapper = (value, failure) => {
			try {
				return Promise.resolve(mapper(value, failure));
			} catch (e) {
				if (e instanceof TerminalError) return Promise.reject(e);
				else {
					ctx.handleInvocationEndError(e);
					return pendingPromise();
				}
			}
		};
	}
	async tryComplete() {
		await this.inner.tryComplete();
	}
	uncompletedLeaves() {
		return this.inner.uncompletedLeaves();
	}
	publicPromise() {
		const promiseMapper = this.publicPromiseMapper;
		return this.inner.publicPromise().then((t) => promiseMapper(t, void 0), (error) => {
			if (error instanceof RestateError) return promiseMapper(void 0, error);
			else throw error;
		});
	}
	[Symbol.toStringTag] = "RestateMappedPromise";
};
/**
* Promises executor, gluing VM with I/O and Promises given to user space.
*/
var PromisesExecutor = class {
	constructor(coreVm, inputPump, outputPump, runClosuresTracker, errorCallback) {
		this.coreVm = coreVm;
		this.inputPump = inputPump;
		this.outputPump = outputPump;
		this.runClosuresTracker = runClosuresTracker;
		this.errorCallback = errorCallback;
	}
	async doProgress(restatePromise) {
		await this.outputPump.awaitNextProgress();
		await this.doProgressInner(restatePromise);
	}
	async doProgressInner(restatePromise) {
		try {
			await restatePromise.tryComplete();
		} catch (e) {
			restatePromise[RESTATE_CTX_SYMBOL].handleInvocationEndError(e);
			return Promise.resolve();
		}
		return setImmediate().then(async () => {
			try {
				const handles = restatePromise.uncompletedLeaves();
				if (handles.length === 0) return;
				const doProgressResult = this.coreVm.do_progress(new Uint32Array(handles));
				if (doProgressResult === "AnyCompleted") {} else if (doProgressResult === "ReadFromInput") await this.inputPump.awaitNextProgress();
				else if (doProgressResult === "WaitingPendingRun") await this.runClosuresTracker.awaitNextCompletedRun();
				else if (doProgressResult === "CancelSignalReceived") {
					restatePromise.tryCancel();
					return;
				} else {
					this.runClosuresTracker.executeRun(doProgressResult.ExecuteRun);
					await setImmediate();
				}
				await this.doProgressInner(restatePromise);
			} catch (e) {
				this.errorCallback(e);
			}
		});
	}
};

//#endregion
export { InvocationPendingPromise, PromisesExecutor, RestateCombinatorPromise, RestateInvocationPromise, RestatePendingPromise, RestateSinglePromise, extractContext, pendingPromise };
