// @restatedev/restate-sdk/dist/logging/console_logger_transport.cjs
const require_logger_transport = require('./logger_transport.cjs');

//#region src/logging/console_logger_transport.ts
const defaultLoggerTransport = (params, message, ...optionalParams) => {
	if (params.replaying) return;
	if (logLevel(params.level) < logLevel(DEFAULT_CONSOLE_LOGGER_LOG_LEVEL)) return;
	const p = `${formatLogPrefix(params.context)} ${params.level.toUpperCase()}:`;
	switch (params.level) {
		case require_logger_transport.RestateLogLevel.TRACE: return console.trace(p, message, ...optionalParams);
		case require_logger_transport.RestateLogLevel.DEBUG: return console.debug(p, message, ...optionalParams);
		case require_logger_transport.RestateLogLevel.INFO: return console.info(p, message, ...optionalParams);
		case require_logger_transport.RestateLogLevel.WARN: return console.warn(p, message, ...optionalParams);
		case require_logger_transport.RestateLogLevel.ERROR: return console.error(p, message, ...optionalParams);
		default: throw new TypeError(`unset or unknown log level ${params.level}`);
	}
};
const DEFAULT_CONSOLE_LOGGER_LOG_LEVEL = readRestateLogLevel();
function readRestateLogLevel() {
	const env = globalThis.process?.env?.RESTATE_LOGGING;
	const level = logLevelFromName(env);
	if (level !== null) return level;
	return require_logger_transport.RestateLogLevel.INFO;
}
function logLevelFromName(name) {
	if (!name) return null;
	switch (name.toUpperCase()) {
		case "TRACE": return require_logger_transport.RestateLogLevel.TRACE;
		case "DEBUG": return require_logger_transport.RestateLogLevel.DEBUG;
		case "INFO": return require_logger_transport.RestateLogLevel.INFO;
		case "WARN": return require_logger_transport.RestateLogLevel.WARN;
		case "ERROR": return require_logger_transport.RestateLogLevel.ERROR;
		default: throw new TypeError(`unknown name ${name}`);
	}
}
function logLevel(level) {
	switch (level) {
		case require_logger_transport.RestateLogLevel.TRACE: return 1;
		case require_logger_transport.RestateLogLevel.DEBUG: return 2;
		case require_logger_transport.RestateLogLevel.INFO: return 3;
		case require_logger_transport.RestateLogLevel.WARN: return 4;
		case require_logger_transport.RestateLogLevel.ERROR: return 5;
	}
}
function formatLogPrefix(context) {
	let prefix = `[restate][${(/* @__PURE__ */ new Date()).toISOString()}]`;
	if (context === void 0) return prefix;
	prefix = `${prefix}[${context.invocationTarget}][${context.invocationId}]`;
	if (context.additionalContext !== void 0) for (const [k, v] of Object.entries(context.additionalContext)) prefix = prefix + `[${k}: ${v}]`;
	return prefix;
}

//#endregion
exports.DEFAULT_CONSOLE_LOGGER_LOG_LEVEL = DEFAULT_CONSOLE_LOGGER_LOG_LEVEL;
exports.defaultLoggerTransport = defaultLoggerTransport;
