// @restatedev/restate-sdk/dist/logging/logger.cjs
const require_logger_transport = require('./logger_transport.cjs');

//#region src/logging/logger.ts
function createLogger(loggerTransport, source, context, isReplaying = () => false) {
	function loggerForLevel(loggerTransport$1, source$1, level, isReplaying$1, context$1) {
		return { get: () => {
			return loggerTransport$1.bind(null, {
				source: source$1,
				level,
				replaying: isReplaying$1(),
				context: context$1
			});
		} };
	}
	const info = loggerForLevel(loggerTransport, source, require_logger_transport.RestateLogLevel.INFO, isReplaying, context);
	return Object.create(console, {
		trace: loggerForLevel(loggerTransport, source, require_logger_transport.RestateLogLevel.TRACE, isReplaying, context),
		debug: loggerForLevel(loggerTransport, source, require_logger_transport.RestateLogLevel.DEBUG, isReplaying, context),
		info,
		log: info,
		warn: loggerForLevel(loggerTransport, source, require_logger_transport.RestateLogLevel.WARN, isReplaying, context),
		error: loggerForLevel(loggerTransport, source, require_logger_transport.RestateLogLevel.ERROR, isReplaying, context),
		logForLevel: { get() {
			return (level, message, ...optionalParams) => {
				if (optionalParams?.length === 0 || optionalParams === void 0) loggerTransport.bind(null)({
					source,
					level,
					replaying: isReplaying(),
					context
				}, message);
				else loggerTransport.bind(null)({
					source,
					level,
					replaying: isReplaying(),
					context
				}, message, optionalParams);
			};
		} }
	});
}

//#endregion
exports.createLogger = createLogger;
