// @restatedev/restate-sdk/dist/logging/logger_transport.js
//#region src/logging/logger_transport.ts
/**
* Logger level.
*/
let RestateLogLevel = /* @__PURE__ */ function(RestateLogLevel$1) {
	RestateLogLevel$1["TRACE"] = "trace";
	RestateLogLevel$1["DEBUG"] = "debug";
	RestateLogLevel$1["INFO"] = "info";
	RestateLogLevel$1["WARN"] = "warn";
	RestateLogLevel$1["ERROR"] = "error";
	return RestateLogLevel$1;
}({});
/**
* Source of the log.
*/
let LogSource = /* @__PURE__ */ function(LogSource$1) {
	LogSource$1["SYSTEM"] = "SYSTEM";
	LogSource$1["JOURNAL"] = "JOURNAL";
	LogSource$1["USER"] = "USER";
	return LogSource$1;
}({});
/**
* Logger context.
*/
var LoggerContext = class {
	invocationTarget;
	constructor(invocationId, serviceName, handlerName, key, request, additionalContext) {
		this.invocationId = invocationId;
		this.serviceName = serviceName;
		this.handlerName = handlerName;
		this.key = key;
		this.request = request;
		this.additionalContext = additionalContext;
		this.invocationTarget = key === void 0 || key.length === 0 ? `${serviceName}/${handlerName}` : `${serviceName}/${key}/${handlerName}`;
	}
};

//#endregion
export { LogSource, LoggerContext, RestateLogLevel };
