// @restatedev/restate-sdk/dist/endpoint/handlers/lambda.cjs
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.cjs');
const require_errors = require('../../types/errors.cjs');
const require_user_agent = require('../../user_agent.cjs');
const require_streams = require('../../utils/streams.cjs');
const require_generic = require('./generic.cjs');
let node_stream_web = require("node:stream/web");
node_stream_web = require_rolldown_runtime.__toESM(node_stream_web);
let node_buffer = require("node:buffer");
node_buffer = require_rolldown_runtime.__toESM(node_buffer);
let node_zlib = require("node:zlib");
node_zlib = require_rolldown_runtime.__toESM(node_zlib);

//#region src/endpoint/handlers/lambda.ts
const RESPONSE_COMPRESSION_THRESHOLD = 3 * 1024 * 1024;
var LambdaHandler = class {
	constructor(handler, compressionSupported) {
		this.handler = handler;
		this.compressionSupported = compressionSupported;
	}
	async handleRequest(event, context) {
		const path = "path" in event ? event.path : event.rawPath;
		let requestContentEncoding;
		let requestAcceptEncoding;
		for (const [key, value] of Object.entries(event.headers)) if (key.localeCompare("content-encoding", void 0, { sensitivity: "accent" }) === 0) requestContentEncoding = value;
		else if (key.localeCompare("accept-encoding", void 0, { sensitivity: "accent" }) === 0) requestAcceptEncoding = value;
		let bodyStream;
		if (!event.body) bodyStream = null;
		else {
			let bodyBuffer;
			if (event.isBase64Encoded) bodyBuffer = node_buffer.Buffer.from(event.body, "base64");
			else bodyBuffer = node_buffer.Buffer.from(new TextEncoder().encode(event.body));
			if (requestContentEncoding && requestContentEncoding.includes("zstd")) {
				if (!this.compressionSupported) throw new Error("The input is compressed using zstd, but this lambda deployment doesn't support compression. Make sure to deploy the Lambda using Node > 22");
				bodyBuffer = node_zlib.zstdDecompressSync(bodyBuffer);
			}
			bodyStream = require_streams.OnceStream(bodyBuffer);
		}
		const abortController = new AbortController();
		const request = {
			body: bodyStream,
			headers: event.headers,
			url: path,
			extraArgs: [context],
			abortSignal: abortController.signal
		};
		let response;
		try {
			response = await this.handler.handle(request, { AWSRequestId: context.awsRequestId });
		} catch (e) {
			abortController.abort();
			throw e;
		}
		const chunks = [];
		try {
			await response.body.pipeTo(new node_stream_web.WritableStream({ write: (chunk) => {
				chunks.push(chunk);
			} }));
		} catch (e) {
			const error = require_errors.ensureError(e);
			(require_generic.tryCreateContextualLogger(this.handler.endpoint.loggerTransport, request.url, request.headers) ?? this.handler.endpoint.rlog).error("Error while collecting invocation response: " + (error.stack ?? error.message));
			return {
				headers: {
					"content-type": "application/json",
					"x-restate-server": require_user_agent.X_RESTATE_SERVER
				},
				statusCode: 500,
				isBase64Encoded: false,
				body: JSON.stringify({ message: error.message })
			};
		} finally {
			abortController.abort();
		}
		const responseBodyBuffer = node_buffer.Buffer.concat(chunks);
		let responseBody;
		if (this.compressionSupported && responseBodyBuffer.length > RESPONSE_COMPRESSION_THRESHOLD && requestAcceptEncoding && requestAcceptEncoding.includes("zstd")) {
			response.headers["content-encoding"] = "zstd";
			responseBody = node_zlib.zstdCompressSync(responseBodyBuffer).toString("base64");
		} else responseBody = responseBodyBuffer.toString("base64");
		return {
			headers: response.headers,
			statusCode: response.statusCode,
			isBase64Encoded: true,
			body: responseBody
		};
	}
};
function isCompressionSupported() {
	return "zstdDecompressSync" in node_zlib && "zstdCompressSync" in node_zlib;
}

//#endregion
exports.LambdaHandler = LambdaHandler;
exports.isCompressionSupported = isCompressionSupported;
