// @restatedev/restate-sdk/dist/endpoint/handlers/lambda.js
import { ensureError } from "../../types/errors.js";
import { X_RESTATE_SERVER } from "../../user_agent.js";
import { OnceStream } from "../../utils/streams.js";
import { tryCreateContextualLogger } from "./generic.js";
import { WritableStream } from "node:stream/web";
import { Buffer } from "node:buffer";
import * as zlib from "node:zlib";

//#region src/endpoint/handlers/lambda.ts
const RESPONSE_COMPRESSION_THRESHOLD = 3 * 1024 * 1024;
var LambdaHandler = class {
	constructor(handler, compressionSupported) {
		this.handler = handler;
		this.compressionSupported = compressionSupported;
	}
	async handleRequest(event, context) {
		const path = "path" in event ? event.path : event.rawPath;
		let requestContentEncoding;
		let requestAcceptEncoding;
		for (const [key, value] of Object.entries(event.headers)) if (key.localeCompare("content-encoding", void 0, { sensitivity: "accent" }) === 0) requestContentEncoding = value;
		else if (key.localeCompare("accept-encoding", void 0, { sensitivity: "accent" }) === 0) requestAcceptEncoding = value;
		let bodyStream;
		if (!event.body) bodyStream = null;
		else {
			let bodyBuffer;
			if (event.isBase64Encoded) bodyBuffer = Buffer.from(event.body, "base64");
			else bodyBuffer = Buffer.from(new TextEncoder().encode(event.body));
			if (requestContentEncoding && requestContentEncoding.includes("zstd")) {
				if (!this.compressionSupported) throw new Error("The input is compressed using zstd, but this lambda deployment doesn't support compression. Make sure to deploy the Lambda using Node > 22");
				bodyBuffer = zlib.zstdDecompressSync(bodyBuffer);
			}
			bodyStream = OnceStream(bodyBuffer);
		}
		const abortController = new AbortController();
		const request = {
			body: bodyStream,
			headers: event.headers,
			url: path,
			extraArgs: [context],
			abortSignal: abortController.signal
		};
		let response;
		try {
			response = await this.handler.handle(request, { AWSRequestId: context.awsRequestId });
		} catch (e) {
			abortController.abort();
			throw e;
		}
		const chunks = [];
		try {
			await response.body.pipeTo(new WritableStream({ write: (chunk) => {
				chunks.push(chunk);
			} }));
		} catch (e) {
			const error = ensureError(e);
			(tryCreateContextualLogger(this.handler.endpoint.loggerTransport, request.url, request.headers) ?? this.handler.endpoint.rlog).error("Error while collecting invocation response: " + (error.stack ?? error.message));
			return {
				headers: {
					"content-type": "application/json",
					"x-restate-server": X_RESTATE_SERVER
				},
				statusCode: 500,
				isBase64Encoded: false,
				body: JSON.stringify({ message: error.message })
			};
		} finally {
			abortController.abort();
		}
		const responseBodyBuffer = Buffer.concat(chunks);
		let responseBody;
		if (this.compressionSupported && responseBodyBuffer.length > RESPONSE_COMPRESSION_THRESHOLD && requestAcceptEncoding && requestAcceptEncoding.includes("zstd")) {
			response.headers["content-encoding"] = "zstd";
			responseBody = zlib.zstdCompressSync(responseBodyBuffer).toString("base64");
		} else responseBody = responseBodyBuffer.toString("base64");
		return {
			headers: response.headers,
			statusCode: response.statusCode,
			isBase64Encoded: true,
			body: responseBody
		};
	}
};
function isCompressionSupported() {
	return "zstdDecompressSync" in zlib && "zstdCompressSync" in zlib;
}

//#endregion
export { LambdaHandler, isCompressionSupported };
